### ISSUE TITLE
Intermediate outputs of pipeline

### ISSUE THREAD

**Issue by AlexanderGeiger** [D3-AI/Orion#61]
### Description

We want to store intermediate outputs of the pipeline. The user might want to store more information about the datarun besides just the found anomalies. 

### Approach

We add another collection `IO` to the DB, which has the fields `datarun` and `io`. `datarun` is a reference to the datarun that created the intermediate output and `io` is a field that stores a dictionary.
A user then could use the pipeline JSON and specify what to store, following the specifications introduced to MLBlocks [here](https://github.com/HDI-Project/MLBlocks/issues/104).

Once we have that feature of MLBlocks, we can make use of the intermediate outputs and store them as a dictionary, using primitive.variable_name as key.

The anomalies are stored in the events collection just like before, therefore those should be returned normally and not as intermediate outputs.

**Comment by dyuliu** [comment]
I prefer to create a new database to store more information that could possibly very large in size. We can still add collection `IO` to the Orion DB. The new collection contains the fields `datarun` and `io`, where `datarun` is the reference to the datarun that created the outputs and `io` is a field that stores a dictionary. 

The different thing here is that `io` stores multiple pairs of `key: value`, where a key is "primitive.variable_name" and "value" is a reference id to the new database (e.g., orionExt). With the reference id, we can retrieve the relevant document/documents in the new database. We can also retrieve back for a given document in the new database to locate which datarun it is related to.

There are two benefits of doing so:
(1) keep the size of the original DB as small as possible, only storing the essential information.
(2) avoid creating a document of too large size (MongoDB suggests doc size <= 16MB). If we append every extra information (e.g., the time-series after aggregation) to `io` field, the document size will increase fast.

Basically, the purpose of dumping extra information is for later analysis (e.g., used in MTV). The extra information usually needs to be further processed to new formats. Let's take Orion + MTV as an example, we now have the following workflow:

- Specify a pipeline, and a dataset (signal set). 
- Run ORION
- It creates one DB#1 per the schema specified in ORION, as well as the other DB#2 storing intermediate outputs in certain efficient schema.
- Dump DB#1 and DB#2, and create a new DB like "MTV" for MTV use  
- Restore DB#1 to "MTV" without any changes
- Get data from DB#2, process it to the desired formats, and store the processed data to "MTV"
- Open MTV app to do analysis

**Comment by csala** [comment]
One of the things that @dyuliu said is crucial: MongoDB has a maximum document size of 16MB, which means that if we store all the values as a document field we might hit size issues.

However, I don't understand the steps that you specified at the end, especially the dump/restore cycle. Would you mind explaining it @dyuliu ?

In any case, I would like to suggest a different alternative: What about using GridFS to store the values?

My suggestion is as follows: For each datarun, we capture the intermediate outputs in a dictionary and then store that dictionary as a pickle file in GridFS within the same MongoDB database.
The name of the file would be `<datarun-id>.pkl`. Also, once MTV preprocesses the data in whatever format needed, the results can also be stored back to GridFS, either as additional fields within the same pickle file or in another pickle file `<datarun-id>.mtv.pkl` or similar.

This will:
- Solve the document size problem: GridFS can store files or arbitrary size.
- Allow any kind of data to be stored, as dataframes or numpy arrays might not be able to be stored in a regular MongoDB document unless serialized first.
- Allow pulling all the data from a Datarun in one shot already in python binary format.

Another alternative in case MTV might need to pull variables individually, would be to store each variable in a separated file: `<datarun_id>.<variable_name>.pkl`.

**Comment by dyuliu** [comment]
@csala The idea of using GridFS is good.

Dump/restore means the MongoDB's commands `mongodump` and `mongorestore`.
[https://docs.mongodb.com/manual/reference/program/mongodump/](url)
[https://docs.mongodb.com/manual/reference/program/mongorestore/](url)

`mongodump` can dump a database to a set of BSON files and store the files in our file system.
`mongorestore` can quickly restore these BSON files to a database in MongoDB.

The basic idea here is similar to yours, saving pickle file in GridFS and retore the data from GridFS for whatever use. As you said, using GridFS has one prominent benefit, that is, storing any kind of data that is not able to be saved as a regular MongoDB document.

**Comment by csala** [comment]
> Dump/restore means the MongoDB's commands `mongodump` and `mongorestore`.
> [https://docs.mongodb.com/manual/reference/program/mongodump/](url)
> [https://docs.mongodb.com/manual/reference/program/mongorestore/](url)
> 
> `mongodump` can dump a database to a set of BSON files and store the files in our file system.
> `mongorestore` can quickly restore these BSON files to a database in MongoDB.

Oh, yes, that was clear already :-)

But what I meant was that I didn't understand why you suggested going through a dump/restore cycle as the regular workflow. But I guess that was because this is the recommended approach now that copyDatabase is deprecated, right?

In any case, to summarize and wrap up the first version, do we agree on the following approach?
- IO data will be stored in GridFS within the same MongoDB Database as Orion.
- All the IO data for a single datarun will be stored as a single dictionary that contains all the data, serialized using pickle.
- The filename will contain the datarun_id. Additionally, I suggest setting the datarun_id within the file metadata, to ease queries.

For an initial version, I suggest storing any preprocessing that MTV needs to perform on the same pickle file (so, load the raw data, preprocess, add the variables to the dictionary, and overwrite the file with it).
If later on we see that the usage requires it, we can decide to split up the file in either raw/preprocessed or even with a single file per variable.
